draft-ietf-tls-esni.md

```diff
@@ -782,10 +777,30 @@ including servers which do not implement this specification.
 
 ### HelloRetryRequest
 
-It is an error for the client to offer ECH before the HelloRetryRequest but not
-after. Likewise, it is an error for the client to offer ECH after the
-HelloRetryRequest but not before. If either of these conditions occurs, then the
-client-facing server MUST abort the handshake with an "illegal_parameter" alert.
+In case a HelloRetryRequest (HRR) is sent, the client-facing server enforces the
+following invariant. Let CH1' denote the first ClientHello processed by the
+server and let CH2' denote the second: either CH1' == CH2' == ClientHelloOuter
+or CH1' == CH2' == ClientHelloInner. In particular, it is an error to process
```

Comment:
I found this a little hard to interpret. It suggests equality of the ClientHellos themselves, which isn't what we want. It's also a little odd since CH1' and CH2' are functions of both what the client sent and what the server chose to do what in response. But it would be less enforcing an invariant on the latter and more maintaining it. WDYT about:

> In case a HelloRetryRequest (HRR) is sent, the client-facing server MUST consistently accept or decline ECH between the two ClientHellos, using the same ECHConfig, and fail the handshake if this is not possible. This is achieved as follows.

Reply:
I like it. Replacing my paragraph with yours. (Changed "fail the handshake" to "abort the handshake".)